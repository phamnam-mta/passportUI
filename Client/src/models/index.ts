export type SplitPaneSizes = {
    analyzeSplitPaneSize: number[];
    labelSplitPaneSize: number[];
    labelTableSplitPaneSize: number[];
};
