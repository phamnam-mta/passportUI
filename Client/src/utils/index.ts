export const delay = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

export const getPixelWidthFromPercent = (widthInPercentage: number): number => {
    const totalWindowWidth = document.documentElement.clientWidth;
    const paneWidthInPixel = (totalWindowWidth / 100) * widthInPercentage;
    return paneWidthInPixel;
};

export const loadCanvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
    return new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            if (blob) {
                resolve(blob);
            } else {
                reject("Failed to create blob for canvas.");
            }
        });
    });
};

export const loadUrlToArrayBuffer = async (url: string): Promise<ArrayBuffer> => {
    return await fetch(url).then((r) => r.arrayBuffer());
};

export const debounce = (func, timeout = 300) => {
    return (...args) => {
        return setTimeout(() => {
            func.apply(this, args);
        }, timeout);
    };
};
