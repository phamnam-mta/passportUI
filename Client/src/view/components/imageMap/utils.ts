import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import ImageLayer from "ol/layer/Image";
import Layer from "ol/layer/Layer";
import VectorLayer from "ol/layer/Vector";
import Projection from "ol/proj/Projection";
import ImageSource from "ol/source/ImageStatic";
import VectorSource from "ol/source/Vector";
import { LAYER_NAME } from "./constants";
import { IRegion } from "./contracts";

export type Dimension = {
    width: number;
    height: number;
};

export const degreeToRadians = (degree: number) => {
    // convert degree to radians
    return (degree * Math.PI * 2) / 360;
};

export const makeLayerFilter = (layerName) => {
    return {
        layerFilter: (layer: Layer) => layer.get(LAYER_NAME) === layerName,
    };
};

export const makeVectorLayer = (layerName, options = {}): Layer => {
    return new VectorLayer({
        [LAYER_NAME]: layerName,
        source: new VectorSource(),
        ...options,
    } as any);
};

export const makeImageLayer = (layerName, uri, projection, imageExtent, options = {}): Layer => {
    return new ImageLayer({
        [LAYER_NAME]: layerName,
        source: makeImageSource(uri, projection, imageExtent),
        ...options,
    } as any);
};

export const makeImageSource = (url: string, projection: Projection, imageExtent: Extent) => {
    return new ImageSource({ url, projection, imageExtent });
};

export const convertRegionToFeature = (region: IRegion, imageExtent: Extent, isOcrProposal: boolean = false) => {
    const coordinates: Array<[number, number]> = [];
    const boundingBox = region.id.split(",").map(parseFloat);
    const imageWidth = imageExtent[2] - imageExtent[0];
    const imageHeight = imageExtent[3] - imageExtent[1];
    for (let i = 0; i < boundingBox.length; i += 2) {
        coordinates.push([Math.round(boundingBox[i] * imageWidth), Math.round((1 - boundingBox[i + 1]) * imageHeight)]);
    }
    const feature = new Feature({
        geometry: new Polygon([coordinates]),
    });
    feature.setProperties({
        id: region.id,
        text: region.value,
        highlighted: false,
        isOcrProposal,
    });
    feature.setId(region.id);

    return feature;
};

export const createRegionIdFromPolygon = (polygon: number[], page: number): string => {
    return polygon.join(",") + ":" + page;
};

export const convertToImageMapCoordinates = (polygon: number[], canvasSize: Dimension, documentSize: Dimension) => {
    const { width: canvasWidth, height: canvasHeight } = canvasSize;
    const { width: ocrWidth, height: ocrHeight } = documentSize;

    const coordinates: number[][] = [];
    for (let i = 0; i < polygon.length; i += 2) {
        coordinates.push([(polygon[i] / ocrWidth) * canvasWidth, (1 - polygon[i + 1] / ocrHeight) * canvasHeight]);
    }
    return coordinates;
};

export const getBoundingBoxFromFeatureId = (id) => id.split(",").map(parseFloat);
