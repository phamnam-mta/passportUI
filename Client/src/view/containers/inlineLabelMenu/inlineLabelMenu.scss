.inline-label-menu-item {
    width: 100%;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.inline-label-menu-input {
    padding: 6px;
}

.inline-label-menu-list {
    overflow: auto;
}
