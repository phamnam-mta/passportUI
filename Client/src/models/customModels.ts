import { FeatureCategory } from "view/components/imageMap/contracts";
import { Polygon } from "./analyzeResult";

export type CustomModel = {
    modelId: string;
    description: string;
    createdDateTime: string;
    apiVersion?: string;
    tags?: {
        modelType?: string;
        [key: string]: string | undefined;
    };
};

export type PrimitiveField = {
    fieldKey: string;
    fieldType: FieldType;
    fieldFormat: FieldFormat;
};

export type ObjectField = {
    fieldKey: string;
    fieldType: FieldType;
    fields: Field[];
    fieldFormat: FieldFormat;
    visualizationHint?: VisualizationHint;
};

export type ArrayField = {
    fieldKey: string;
    fieldType: FieldType;
    itemType: string;
    visualizationHint?: VisualizationHint;
};

export type Field = PrimitiveField | ObjectField | ArrayField;

export type FieldsWithOrder = Field & { order: number };

export enum FieldType {
    String = "string",
    Number = "number",
    Date = "date",
    Time = "time",
    Integer = "integer",
    SelectionMark = "selectionMark",
    Signature = "signature",
    Array = "array",
    Object = "object",
}

export enum FieldFormat {
    NotSpecified = "not-specified",
    Currency = "currency",
    Decimal = "decimal",
    DecimalCommaSeparated = "decimal-comma-separated",
    NoWhiteSpaces = "no-white-spaces",
    Alphanumeric = "alphanumeric",
    DMY = "dmy",
    MDY = "mdy",
    YMD = "ymd",
}

export enum VisualizationHint {
    Horizontal = "horizontal",
    Vertical = "vertical",
}

export type Labels = {
    [documentName: string]: Label[];
};

export type Label = {
    label: string;
    value: LabelValue[];
    labelType?: LabelType;
};

export type LabelValue = {
    boundingBoxes: Polygon[];
    page: number;
    text: string;
};

export type LabelValueCandidate = {
    boundingBoxes: Polygon[];
    page: number;
    text: string;
    category: FeatureCategory;
    alreadyAssignedLabelName?: string;
};

export enum LabelType {
    Words = "words",
    Region = "region",
}

export type Definitions = {
    [objectName: string]: ObjectField;
};

export enum TableType {
    dynamic = "dynamic",
    fixed = "fixed",
}

export enum HeaderType {
    row = "row",
    column = "column",
}
