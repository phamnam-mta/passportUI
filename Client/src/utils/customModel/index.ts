import { SharedColors } from "@fluentui/react";
import { AnalyzeResultAdapterFactory } from "adapters/analyzeResultAdapter";
import { colors } from "consts/colors";
import { Field, FieldType, Label, Labels, LabelValue, LabelValueCandidate } from "models/customModels";
import { CustomModelAssetService } from "services/assetService/customModelAssetService";
import { FieldLocation, ICustomModelError } from "store/customModel/customModel";
import { IDocument } from "store/documents/documentsTypes";
import { FeatureCategory } from "view/components/imageMap/contracts";
import { createRegionIdFromPolygon } from "view/components/imageMap/utils";

export const buildRegionOrders = (analyzeResult) => {
    const analyzeResultAdapter = AnalyzeResultAdapterFactory.create(analyzeResult);
    const documentPage = analyzeResultAdapter.getDocumentPages();
    const regionOrders = {};
    documentPage.forEach((page) => {
        const { pageNumber, selectionMarks, width, height } = page;

        let order = 0;
        page.words.forEach((word) => {
            const { polygon } = word;
            if (polygon) {
                const normalizedPoints: number[] = [];
                for (let i = 0; i < polygon.length; i += 2) {
                    normalizedPoints.push(polygon[i] / width);
                    normalizedPoints.push(polygon[i + 1] / height);
                }
                const orderId = createRegionIdFromPolygon(normalizedPoints, pageNumber);
                regionOrders[orderId] = order++;
            }
        });

        if (selectionMarks) {
            selectionMarks.forEach((selectionMark) => {
                const { polygon } = selectionMark;
                if (polygon) {
                    const normalizedPoints: number[] = [];
                    for (let i = 0; i < polygon.length; i += 2) {
                        normalizedPoints.push(polygon[i] / width);
                        normalizedPoints.push(polygon[i + 1] / height);
                    }
                    const orderId = createRegionIdFromPolygon(normalizedPoints, pageNumber);
                    regionOrders[orderId] = order++;
                }
            });
        }
    });
    return regionOrders;
};

export const decodeLabelString = (label: string): string => {
    return label.replace(/~1/g, "/").replace(/~0/g, "~");
};

export const getFieldKeyFromLabel = (label: Label): string => {
    return decodeLabelString(label.label.split("/")[0]);
};

export const getTableFieldKeyFromLabel = (label: Label, fieldLocation: FieldLocation): string => {
    const labelIndex = fieldLocation === FieldLocation.field ? 1 : 2;
    return decodeLabelString(label.label.split("/")[labelIndex]);
};

export const getAllDocumentLabels = async (
    labels: Labels,
    documents: IDocument[],
    assetService: CustomModelAssetService
) => {
    const documentsWithoutLabels = documents.filter((document) => !labels[document.name]).map((doc) => doc.name);
    const addedLabels = await assetService.fetchAllDocumentLabels(documentsWithoutLabels);
    return { ...labels, ...addedLabels };
};

export const encodeLabelString = (label: string): string => {
    // Encoding for "/" since it will be used in table labeling result.
    return label.replace(/~/g, "~0").replace(/\//g, "~1");
};

export const getColorByFieldKey = (colorForFields: Record<string, string>[], fieldKey: string) => {
    const getDynamicKey = (obj: Record<string, string>) => Object.keys(obj)[0];

    const colorMap = colorForFields.find((color) => getDynamicKey(color) === fieldKey);
    return colorMap ? colorMap[fieldKey] : SharedColors.gray10;
};

export const getFieldColor = (fields: Field[], key: string): string => {
    const index = fields.findIndex((field) => field.fieldKey === key);
    const colorCounts = colors.length;

    if (index !== -1) {
        return colors[index % colorCounts];
    } else {
        return SharedColors.gray10;
    }
};

export const uniqueByKeepFirst = (array, key) => {
    let seen = new Set();
    return array.filter((item) => {
        let k = key(item);
        return seen.has(k) ? false : seen.add(k);
    });
};

export const makeLabelValue = (labelValueCandidate: LabelValueCandidate) => {
    const { page, text, boundingBoxes } = labelValueCandidate;
    return { page, text, boundingBoxes };
};

export const getOrder = (currentOrders: { [key: string]: number }, orderId: string): number => {
    const orderNumber = currentOrders[orderId];
    if (orderNumber !== undefined) {
        return orderNumber;
    }
    return 0;
};

export const compareOrder = (a: LabelValue, b: LabelValue, currentOrders, currentPage) => {
    const order1 = getOrder(currentOrders, createRegionIdFromPolygon(a.boundingBoxes[0], currentPage));
    const order2 = getOrder(currentOrders, createRegionIdFromPolygon(b.boundingBoxes[0], currentPage));
    return order1 >= order2 ? 1 : -1;
};

export const makeError = (name: string, message: string): ICustomModelError => {
    return { name, message };
};

export const validateAssignment = (candidates: LabelValueCandidate[], field: Field) => {
    switch (field.fieldType) {
        case FieldType.Signature:
            if (candidates.length !== 1 || candidates[0].category !== FeatureCategory.DrawnRegion) {
                throw makeError("Label assignment warning", "Signature field only supports one draw region per field.");
            }
            break;
        case FieldType.SelectionMark:
            if (
                candidates.length !== 1 ||
                !(
                    candidates[0].category === FeatureCategory.Checkbox ||
                    candidates[0].category === FeatureCategory.DrawnRegion
                )
            ) {
                throw makeError(
                    "Label assignment warning",
                    "SelectionMark field only supports one draw region or checkbox per field."
                );
            }
            break;
        default:
            // Text or table cell
            if (candidates.length > 1 && candidates.some((c) => c.category === FeatureCategory.DrawnRegion)) {
                throw makeError(
                    "Label assignment warning",
                    "General text field only supports one single draw region, or one or more text boxes."
                );
            }
            break;
    }
};

export const getUnusedFieldColor = (colorForFields: Record<string, string>[]) => {
    const getDynamicValue = (obj: Record<string, string>) => Object.values(obj)[0];

    const colorCounts = colors.length;
    const usedColors = colorForFields.map((field) => getDynamicValue(field));
    const unusedColors = colors.filter((color) => !usedColors.includes(color));

    if (colorForFields.length >= colorCounts && unusedColors.length === 0) {
        const initialUsageCount = colors.reduce((acc, color) => ({ ...acc, [color]: 0 }), {});
        const usageCount = usedColors.reduce((result, value) => ({ ...result, [value]: (result[value] || 0) + 1 }), {});

        const totalUsageCount = { ...initialUsageCount, ...usageCount };
        const colorCycleRound = Math.ceil(colorForFields.length / colorCounts);

        const leastUsedColorIndex = (Object.values(totalUsageCount) as number[]).findIndex(
            (count: number) => count < colorCycleRound
        );

        return colors[leastUsedColorIndex];
    }

    return unusedColors[0];
};
