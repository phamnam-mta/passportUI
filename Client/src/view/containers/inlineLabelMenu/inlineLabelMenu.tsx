import * as React from "react";
import {
    ActionButton,
    FocusZone,
    FocusZoneDirection,
    List,
    Stack,
    TextField,
    Callout,
    Depths,
    NeutralColors,
} from "@fluentui/react";
import { connect, ConnectedProps } from "react-redux";
import { getColorByFieldKey, encodeLabelString } from "utils/customModel";
import { FieldFormat, FieldType, PrimitiveField } from "models/customModels";
import { ApplicationState } from "store";
import { addField, assignLabel } from "store/customModel/customModel";

import "./inlineLabelMenu.scss";

type MenuItem = {
    iconName: string;
    disabled?: boolean;
    iconColor?: string;
    text: string;
    onClick: () => void;
};

interface IInlineLabelMenuState {
    items: MenuItem[];
    searchText: string;
}

interface IInlineLabelMenuProps {
    showPopup: boolean;
    positionTop: number;
    positionLeft: number;
    enabledTypes: FieldType[];
}

type InlineLabelMenuProps = ConnectedProps<typeof connector> & IInlineLabelMenuProps;

export const inlineLabelMenuHeight = 180;

export class InlineLabelMenu extends React.PureComponent<InlineLabelMenuProps, IInlineLabelMenuState> {
    constructor(props) {
        super(props);
        this.state = {
            items: [],
            searchText: "",
        };
    }

    public componentDidMount() {
        this.prepareItems();
    }

    public componentDidUpdate(prevProps, prevState) {
        if (
            prevState.searchText !== this.state.searchText ||
            prevProps.fields !== this.props.fields ||
            prevProps.colorForFields !== this.props.colorForFields ||
            prevProps.enabledTypes !== this.props.enabledTypes
        ) {
            this.prepareItems();
        }
    }
    private prepareItems = () => {
        const { fields, colorForFields } = this.props;
        const { enabledTypes } = this.props;
        const lowerSearchText = this.state.searchText.toLocaleLowerCase();
        const items = fields
            .filter((f) => f.fieldType !== FieldType.Array && f.fieldType !== FieldType.Object)
            .filter((f) => f.fieldKey.toLocaleLowerCase().includes(lowerSearchText))
            .map((f) => {
                return {
                    iconName: "CircleFill",
                    iconColor: getColorByFieldKey(colorForFields, f.fieldKey),
                    text: f.fieldKey,
                    type: f.fieldType,
                    onClick: () => this.handleFieldClick(f.fieldKey),
                };
            })
            .filter((f) => enabledTypes.includes(f.type));
        this.setState({ items: items.length > 0 ? items : this.makeCreateFieldItems() });
    };

    private makeCreateFieldItems = (): MenuItem[] => [
        {
            text: "Field",
            iconName: "GroupList",
            onClick: () => this.handleCreateFieldClick(this.state.searchText, FieldType.String),
        },
        {
            text: "Select a subscription",
            iconName: "CheckboxComposite",
            onClick: () => this.handleCreateFieldClick(this.state.searchText, FieldType.SelectionMark),
        },
        {
            text: "Signature",
            iconName: "WhiteBoardApp16",
            onClick: () => this.handleCreateFieldClick(this.state.searchText, FieldType.Signature),
        },
    ];

    private handleSearchTextChange = (_, newText?: string) => {
        this.setState({ searchText: newText || "" });
    };

    private handleFieldClick = (fieldKey) => {
        this.props.assignLabel(encodeLabelString(fieldKey));
        this.setState({ searchText: "" });
    };

    private handleCreateFieldClick = async (fieldKey, fieldType) => {
        if (!fieldKey) {
            return; // Ignore if fieldKey is empty;
        }
        const { addField, assignLabel } = this.props;
        const field: PrimitiveField = { fieldKey, fieldType, fieldFormat: FieldFormat.NotSpecified };
        await addField(field);
        assignLabel(encodeLabelString(fieldKey));
        this.setState({ searchText: "" });
    };

    private renderItems = (item?: MenuItem) => {
        return (
            <ActionButton
                className="inline-label-menu-item"
                disabled={item?.disabled}
                styles={{ rootHovered: { backgroundColor: NeutralColors.gray10 } }}
                iconProps={{ iconName: item?.iconName, style: { color: item?.iconColor } }}
                ariaLabel={item?.text}
                onClick={item?.onClick}
            >
                {item?.text}
            </ActionButton>
        );
    };

    public render() {
        const { showPopup, positionTop, positionLeft, hideInlineLabelMenu } = this.props;
        const { items, searchText } = this.state;
        return (
            <>
                {showPopup && !hideInlineLabelMenu && (
                    <Callout
                        styles={{
                            calloutMain: {
                                display: "flex",
                                width: 250,
                                flexDirection: "column",
                                top: positionTop,
                                left: positionLeft,
                                position: "absolute",
                                boxShadow: Depths.depth16,
                            },
                        }}
                        calloutMaxHeight={inlineLabelMenuHeight}
                        isBeakVisible={false}
                        setInitialFocus
                    >
                        <Stack className="inline-label-menu-input">
                            <TextField
                                value={searchText}
                                placeholder="Search existing or create new"
                                autoComplete="off"
                                onChange={this.handleSearchTextChange}
                            />
                        </Stack>
                        <Stack className="inline-label-menu-list" grow={1}>
                            <FocusZone direction={FocusZoneDirection.vertical}>
                                <List items={items} onRenderCell={this.renderItems} />
                            </FocusZone>
                        </Stack>
                    </Callout>
                )}
            </>
        );
    }
}

const mapState = (state: ApplicationState) => ({
    fields: state.customModel.fields,
    colorForFields: state.customModel.colorForFields,
    hideInlineLabelMenu: state.customModel.hideInlineLabelMenu,
});
const mapDispatch = { addField, assignLabel };

const connector = connect(mapState, mapDispatch);

export default connector(InlineLabelMenu);
