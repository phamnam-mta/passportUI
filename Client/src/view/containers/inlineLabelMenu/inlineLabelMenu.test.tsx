import * as React from "react";
import { encodeLabelString } from "utils/customModel";
import { FieldFormat, FieldType } from "models/customModels";
import { flushPromises, mockFields, mockColorForFields } from "utils/test";
import { InlineLabelMenu } from "./inlineLabelMenu";
import { shallow } from "enzyme";

describe("<InlineLabelMenu />", () => {
    let baseProps;
    const mockSearchText = mockFields[0].fieldKey;
    const mockNewFieldText = "mockNewFieldText";

    beforeEach(() => {
        baseProps = {
            addField: jest.fn(),
            assignLabel: jest.fn(),
            showPopup: true,
            positionTop: 200,
            positionLeft: 300,
            fields: mockFields,
            hideInlineLabelMenu: false,
            colorForFields: mockColorForFields,
            enabledTypes: [],
        };
    });

    describe("Rendering", () => {
        it("should render correctly with field items", () => {
            const wrapper = shallow(<InlineLabelMenu {...baseProps} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly with hideInlineLabelMenu is true", () => {
            const props = {
                ...baseProps,
                hideInlineLabelMenu: true,
            };
            const wrapper = shallow(<InlineLabelMenu {...props} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly with filtered items", () => {
            const wrapper = shallow(<InlineLabelMenu {...baseProps} />);
            wrapper.setState({ searchText: mockSearchText });
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly with filtered items by selected text field", () => {
            const props = {
                ...baseProps,
                enabledTypes: [FieldType.String, FieldType.Date, FieldType.Time, FieldType.Integer, FieldType.Number],
            };

            const wrapper = shallow(<InlineLabelMenu {...props} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly with filtered items by selected checkbox field", () => {
            const props = {
                ...baseProps,
                enabledTypes: [FieldType.SelectionMark],
            };

            const wrapper = shallow(<InlineLabelMenu {...props} />);
            expect(wrapper).toMatchSnapshot();
        });

        it("should render correctly with create-field items", () => {
            const wrapper = shallow(<InlineLabelMenu {...baseProps} />);
            wrapper.setState({ searchText: mockNewFieldText });
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should assignLabel when item is clicked", async () => {
            const props = {
                ...baseProps,
                enabledTypes: [FieldType.String],
            };
            const wrapper = shallow(<InlineLabelMenu {...props} />);
            const list = wrapper.find("List");

            const items = list.prop("items") as any;
            items[0].onClick();

            await flushPromises();

            expect(baseProps.assignLabel).toBeCalledTimes(1);
            expect(baseProps.assignLabel).toBeCalledWith(encodeLabelString(items[0].text));
            expect(wrapper.state("searchText")).toBe("");
        });

        it("should skip add field when searchText is empty", async () => {
            const props = { ...baseProps, fields: [] };
            const wrapper = shallow(<InlineLabelMenu {...props} />);
            wrapper.setState({ searchText: "" });

            const stringItem = (wrapper.find("List").prop("items") as any)[0];
            stringItem.onClick();

            await flushPromises();

            expect(baseProps.addField).not.toBeCalled();
        });

        it("should add string field when item is clicked", async () => {
            const props = {
                ...baseProps,
                enabledTypes: [FieldType.String, FieldType.Date, FieldType.Time, FieldType.Integer, FieldType.Number],
            };
            const wrapper = shallow(<InlineLabelMenu {...props} />);
            wrapper.setState({ searchText: mockNewFieldText });

            const stringItem = (wrapper.find("List").prop("items") as any)[0];
            stringItem.onClick();

            await flushPromises();

            expect(baseProps.addField).toBeCalledTimes(1);

            const expectedField = {
                fieldKey: mockNewFieldText,
                fieldType: FieldType.String,
                fieldFormat: FieldFormat.NotSpecified,
            };
            expect(baseProps.addField).toBeCalledWith(expectedField);
            expect(wrapper.state("searchText")).toBe("");
        });

        it("should add selection mark field when item is clicked", async () => {
            const props = {
                ...baseProps,
                enabledTypes: [FieldType.SelectionMark],
            };
            const wrapper = shallow(<InlineLabelMenu {...props} />);
            wrapper.setState({ searchText: mockNewFieldText });

            const selectionMarkItem = (wrapper.find("List").prop("items") as any)[1];
            selectionMarkItem.onClick();

            await flushPromises();

            expect(baseProps.addField).toBeCalledTimes(1);

            const expectedField = {
                fieldKey: mockNewFieldText,
                fieldType: FieldType.SelectionMark,
                fieldFormat: FieldFormat.NotSpecified,
            };
            expect(baseProps.addField).toBeCalledWith(expectedField);
            expect(wrapper.state("searchText")).toBe("");
        });

        it("should set searchText when typing on TextField.", async () => {
            const wrapper = shallow(<InlineLabelMenu {...baseProps} />);

            const textField = wrapper.find("StyledTextFieldBase") as any;
            textField.prop("onChange")(null, mockSearchText);

            expect(wrapper.state("searchText")).toBe(mockSearchText);
        });
    });
});
