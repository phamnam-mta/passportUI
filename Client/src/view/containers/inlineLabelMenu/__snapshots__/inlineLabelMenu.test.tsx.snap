// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineLabelMenu /> Rendering should render correctly with create-field items 1`] = `
<Fragment>
  <Callout
    calloutMaxHeight={180}
    isBeakVisible={false}
    setInitialFocus={true}
    styles={
      Object {
        "calloutMain": Object {
          "boxShadow": "0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108)",
          "display": "flex",
          "flexDirection": "column",
          "left": 300,
          "position": "absolute",
          "top": 200,
          "width": 250,
        },
      }
    }
  >
    <Stack
      className="inline-label-menu-input"
    >
      <StyledTextFieldBase
        autoComplete="off"
        onChange={[Function]}
        placeholder="Search existing or create new"
        value="mockNewFieldText"
      />
    </Stack>
    <Stack
      className="inline-label-menu-list"
      grow={1}
    >
      <FocusZone
        direction={0}
        isCircularNavigation={false}
        shouldRaiseClicks={true}
      >
        <List
          items={
            Array [
              Object {
                "iconName": "GroupList",
                "onClick": [Function],
                "text": "Field",
              },
              Object {
                "iconName": "CheckboxComposite",
                "onClick": [Function],
                "text": "Select a subscription",
              },
              Object {
                "iconName": "WhiteBoardApp16",
                "onClick": [Function],
                "text": "Signature",
              },
            ]
          }
          onRenderCell={[Function]}
          renderedWindowsAhead={2}
          renderedWindowsBehind={2}
          startIndex={0}
        />
      </FocusZone>
    </Stack>
  </Callout>
</Fragment>
`;

exports[`<InlineLabelMenu /> Rendering should render correctly with field items 1`] = `
<Fragment>
  <Callout
    calloutMaxHeight={180}
    isBeakVisible={false}
    setInitialFocus={true}
    styles={
      Object {
        "calloutMain": Object {
          "boxShadow": "0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108)",
          "display": "flex",
          "flexDirection": "column",
          "left": 300,
          "position": "absolute",
          "top": 200,
          "width": 250,
        },
      }
    }
  >
    <Stack
      className="inline-label-menu-input"
    >
      <StyledTextFieldBase
        autoComplete="off"
        onChange={[Function]}
        placeholder="Search existing or create new"
        value=""
      />
    </Stack>
    <Stack
      className="inline-label-menu-list"
      grow={1}
    >
      <FocusZone
        direction={0}
        isCircularNavigation={false}
        shouldRaiseClicks={true}
      >
        <List
          items={
            Array [
              Object {
                "iconName": "GroupList",
                "onClick": [Function],
                "text": "Field",
              },
              Object {
                "iconName": "CheckboxComposite",
                "onClick": [Function],
                "text": "Select a subscription",
              },
              Object {
                "iconName": "WhiteBoardApp16",
                "onClick": [Function],
                "text": "Signature",
              },
            ]
          }
          onRenderCell={[Function]}
          renderedWindowsAhead={2}
          renderedWindowsBehind={2}
          startIndex={0}
        />
      </FocusZone>
    </Stack>
  </Callout>
</Fragment>
`;

exports[`<InlineLabelMenu /> Rendering should render correctly with filtered items 1`] = `
<Fragment>
  <Callout
    calloutMaxHeight={180}
    isBeakVisible={false}
    setInitialFocus={true}
    styles={
      Object {
        "calloutMain": Object {
          "boxShadow": "0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108)",
          "display": "flex",
          "flexDirection": "column",
          "left": 300,
          "position": "absolute",
          "top": 200,
          "width": 250,
        },
      }
    }
  >
    <Stack
      className="inline-label-menu-input"
    >
      <StyledTextFieldBase
        autoComplete="off"
        onChange={[Function]}
        placeholder="Search existing or create new"
        value="Address"
      />
    </Stack>
    <Stack
      className="inline-label-menu-list"
      grow={1}
    >
      <FocusZone
        direction={0}
        isCircularNavigation={false}
        shouldRaiseClicks={true}
      >
        <List
          items={
            Array [
              Object {
                "iconName": "GroupList",
                "onClick": [Function],
                "text": "Field",
              },
              Object {
                "iconName": "CheckboxComposite",
                "onClick": [Function],
                "text": "Select a subscription",
              },
              Object {
                "iconName": "WhiteBoardApp16",
                "onClick": [Function],
                "text": "Signature",
              },
            ]
          }
          onRenderCell={[Function]}
          renderedWindowsAhead={2}
          renderedWindowsBehind={2}
          startIndex={0}
        />
      </FocusZone>
    </Stack>
  </Callout>
</Fragment>
`;

exports[`<InlineLabelMenu /> Rendering should render correctly with filtered items by selected checkbox field 1`] = `
<Fragment>
  <Callout
    calloutMaxHeight={180}
    isBeakVisible={false}
    setInitialFocus={true}
    styles={
      Object {
        "calloutMain": Object {
          "boxShadow": "0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108)",
          "display": "flex",
          "flexDirection": "column",
          "left": 300,
          "position": "absolute",
          "top": 200,
          "width": 250,
        },
      }
    }
  >
    <Stack
      className="inline-label-menu-input"
    >
      <StyledTextFieldBase
        autoComplete="off"
        onChange={[Function]}
        placeholder="Search existing or create new"
        value=""
      />
    </Stack>
    <Stack
      className="inline-label-menu-list"
      grow={1}
    >
      <FocusZone
        direction={0}
        isCircularNavigation={false}
        shouldRaiseClicks={true}
      >
        <List
          items={
            Array [
              Object {
                "iconColor": "#a0aeb2",
                "iconName": "CircleFill",
                "onClick": [Function],
                "text": "SelectionMark",
                "type": "selectionMark",
              },
            ]
          }
          onRenderCell={[Function]}
          renderedWindowsAhead={2}
          renderedWindowsBehind={2}
          startIndex={0}
        />
      </FocusZone>
    </Stack>
  </Callout>
</Fragment>
`;

exports[`<InlineLabelMenu /> Rendering should render correctly with filtered items by selected text field 1`] = `
<Fragment>
  <Callout
    calloutMaxHeight={180}
    isBeakVisible={false}
    setInitialFocus={true}
    styles={
      Object {
        "calloutMain": Object {
          "boxShadow": "0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108)",
          "display": "flex",
          "flexDirection": "column",
          "left": 300,
          "position": "absolute",
          "top": 200,
          "width": 250,
        },
      }
    }
  >
    <Stack
      className="inline-label-menu-input"
    >
      <StyledTextFieldBase
        autoComplete="off"
        onChange={[Function]}
        placeholder="Search existing or create new"
        value=""
      />
    </Stack>
    <Stack
      className="inline-label-menu-list"
      grow={1}
    >
      <FocusZone
        direction={0}
        isCircularNavigation={false}
        shouldRaiseClicks={true}
      >
        <List
          items={
            Array [
              Object {
                "iconColor": "#a0aeb2",
                "iconName": "CircleFill",
                "onClick": [Function],
                "text": "Address",
                "type": "string",
              },
            ]
          }
          onRenderCell={[Function]}
          renderedWindowsAhead={2}
          renderedWindowsBehind={2}
          startIndex={0}
        />
      </FocusZone>
    </Stack>
  </Callout>
</Fragment>
`;

exports[`<InlineLabelMenu /> Rendering should render correctly with hideInlineLabelMenu is true 1`] = `<Fragment />`;
